agent/src/util/discovery_operator.rs

```diff
@@ -0,0 +1,1411 @@
+                }
+            }
+        }
+    }
+    /// Listens for new discovery responses or a notification to stop connecting to the Discovery Handler
+    #[allow(dead_code)]
+    pub async fn internal_do_discover<'a>(
+        &'a self,
+        kube_interface: Arc<Box<dyn k8s::KubeInterface>>,
+        endpoint: &'a str,
+        dh_details: &'a DiscoveryHandlerDetails,
+        stream: &'a mut dyn StreamingExt,
+    ) -> Result<(), Status> {
+        // clone objects for thread
+        let discovery_operator = Arc::new(self.clone());
+        let stop_discovery_receiver: &mut broadcast::Receiver<()> =
+            &mut dh_details.stop_discovery.subscribe();
+        loop {
+            // Wait for either new discovery results or a message to stop discovery
+            tokio::select! {
+                _ = try_receive(stop_discovery_receiver) => {
+                    trace!("internal_do_discover - received message to stop discovery for endpoint {} serving protocol {}", endpoint, discovery_operator.config.spec.protocol.name);
+                    break;
+                },
+                result = stream.get_message() => {
+                    let message = result?;
+                    if let Some(response) = message {
+                        trace!("internal_do_discover - got discovery results {:?}", response.devices);
+                        self.handle_discovery_results(
+                            kube_interface.clone(),
+                            response.devices,
+                            dh_details.register_request.is_local,
+                            Box::new(DevicePluginBuilder{}),
+                        )
+                        .await
+                        .unwrap();
+                    } else {
+                        error!("internal_do_discover - received result of type None. Should not happen.");
+                        break;
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Sets the connectivity status of a discovery handler. If a discovery handler goes offline, mark_offline_or_deregister_discovery_handler should be used.
+    pub fn set_discovery_handler_connectivity_status(
+        &self,
+        endpoint: &str,
+        connectivity_status: DiscoveryHandlerConnectivityStatus,
+    ) {
+        trace!("set_discovery_handler_connectivity_status - discovery handler at endpoint {} and protocol {} is offline", endpoint, self.config.spec.protocol.name);
+        if let Some(registered_dh_map) = self
+            .discovery_handler_map
+            .lock()
+            .unwrap()
+            .get_mut(&self.config.spec.protocol.name)
+        {
+            if let Some(dh_details) = registered_dh_map.get(endpoint) {
+                let mut dh_details = dh_details.clone();
+                dh_details.connectivity_status = connectivity_status;
+                registered_dh_map.insert(endpoint.to_string(), dh_details);
+            }
+        }
```

Comment:
is there any value in logging the failure cases here?  (**here** in this: `if let Some(...) = ... { ... } else { /*here*/}`)

Reply:
at the least, maybe a comment that describes why we don't care about it?

Reply:
Thank you for bringing this up. The function should never be called for a discovery handler that has a Client. I bubbled up this error in 2f6ad02.